src/Microsoft.DotNet.Wpf/src/PresentationFramework/System/Windows/Controls/PopupControlService.cs

```diff
@@ -1173,19 +1137,400 @@ internal KeyboardFocusChangedEventHandler FocusChangedEventHandler
+            {
+                if (value == null)
+                {
+                    _storage = null;
+                }
+                else if (_storage == null)
+                {
+                    _storage = new WeakReference<T>(value);
+                }
+                else
+                {
+                    _storage.SetTarget(value);
+                }
+            }
+        }
+
+        // A region is convex if every line segment connecting two points of the region lies
+        // within the region.  The convex hull of a set of points is the smallest convex region
+        // that contains the points.  This is just what we need for the safe area of a tooltip
+        // and its owner:  the tooltip should remain open as long as the mouse lies on a line
+        // segment connecting some point in the owner to some point in the tooltip, i.e. as long
+        // as the mouse is in the convex hull of the corners of the owner and tooltip rectangles.
+        //
+        // There are several aspects of this use-case we can exploit.
+        //  * The points come from WM_MOUSEMOVE messages, in the coords of the hwnd's client area.
+        //      This means they are 16-bit integers.  We can compute cross-products using
+        //      integer multiplication without fear of overflow.
+        //  * The convex hull is built from only 8 points - the corners of the two rectangles.
+        //      We can use simple algorithms with low overhead, ignoring their less-than-optimal
+        //      asymptotic cost.
+        //  * The convex hull will have between 4 and 8 edges, at least 4 of which are axis-aligned.
+        //      We can test these edges by simple integer comparison, no multiplications needed.
+        class ConvexHull
+        {
+            internal ConvexHull(PresentationSource source, NativeMethods.RECT rect1, NativeMethods.RECT rect2)
+                : this(source)
+            {
+                PointList points = new PointList();
+                AddPoints(points, rect1);
+                AddPoints(points, rect2);
+                SortPoints(points);
+                BuildHullIncrementally(points);
+            }
+
+            internal ConvexHull(PresentationSource source, NativeMethods.RECT rect)
+                : this(source)
+            {
+                PointList points = new PointList();
+                AddPoints(points, rect, rectIsHull:true);
+                _points = points.ToArray();
+            }
+
+            private ConvexHull(PresentationSource source)
+            {
+                _source = source;
+            }
+
+            // sort by y (and by x among equal y's)
+            private void SortPoints(PointList points)
+            {
+                // insertion sort is good enough.  We're dealing with a small
+                // set of points that are nearly in the right order already.
+                for (int i=0, N=points.Count; i<N; ++i)
```

Comment:
Can we start with 1?

```suggestion
                for (int i=1, N=points.Count; i<N; ++i)
```

Reply:
Yes.